source/sigv4.c

```diff
@@ -793,6 +841,231 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t appendSignedHeaders( CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0, buffRemaining, keyLen = 0, i = 0;
+        char * pBufLoc;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        for( noOfHeaders = 0; noOfHeaders < SIGV4_MAX_HTTP_HEADER_COUNT; noOfHeaders++ )
+        {
+            if( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData != NULL )
+            {
+                sigV4Status = writeSignedHeaderToString( noOfHeaders, canonicalRequest );
+
+                if( sigV4Status != SigV4Success )
+                {
+                    break;
+                }
+            }
+        }
+
+        if( sigV4Status == SigV4Success )
+        {
+            pBufLoc--;
+            *pBufLoc = '\n';
+        }
+
+        return sigV4Status;
+    }
+    static SigV4Status_t writeCanonicalHeaderToString( size_t headerIndex,
+                                                       CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0, buffRemaining = 0, keyLen = 0, valLen = 0, i = 0, trimValueLen = 0;
+        char * pBufLoc;
+        const char * value;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerNum ].key.dataLen;
+        valLen = canonicalRequest->pHeadersLoc[ headerNum ].value.dataLen;
+
+        if( keyLen + valLen + 2 > buffRemaining )
+        {
+            return SigV4InsufficientMemory;
+        }
+        else
+        {
+            for( i = 0; i < keyLen; i++ )
+            {
+                *pBufLoc = tolower( canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ] );
+                pBufLoc++;
+            }
+
+            *pBufLoc = ':';
+            pBufLoc++;
+            value = canonicalRequest->pHeadersLoc[ headerNum ].value.pData;
+            trimValueLen = 0;
+
+            for( i = 0; i < valLen; i++ )
+            {
+                if( isspace( value[ i ] ) && ( i + 1 <= valLen ) && ( ( i + 1 == valLen ) || isspace( value[ i + 1 ] ) || ( trimValueLen == 0 ) ) )
+                {
+                    continue;
+                }
+                else
+                {
+                    *pBufLoc = canonicalRequest->pHeadersLoc[ headerNum ].value.pData[ i ];
+                    pBufLoc++;
+                    trimValueLen++;
+                }
+            }
+
+            *pBufLoc = '\n';
+            pBufLoc++;
+
+            buffRemaining -= ( keyLen + trimValueLen + 2 );
```

Comment:
This looks like it may be susceptible to overflow. CBMC should catch it though.

Reply:
It will not overflow as before writing to buffer I have checked that there should be enough buffer to accommodate the data.`if( keyLen + valLen + 2 > buffRemaining )
        {
            return SigV4InsufficientMemory;
        }`